lib/ddtrace/contrib/grape/endpoint.rb

```diff
@@ -0,0 +1,156 @@
+        KEY_RENDER = 'datadog_grape_endpoint_render'.freeze
+
+        def self.subscribe
+          # Grape is instrumented only if it's available
+          return unless defined?(::Grape) && defined?(::ActiveSupport::Notifications)
+
+          # subscribe when a Grape endpoint is hit
+          ::ActiveSupport::Notifications.subscribe('endpoint_run.grape.start_process') do |*args|
+            endpoint_start_process(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_run.grape') do |*args|
+            endpoint_run(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_render.grape.start_render') do |*args|
+            endpoint_start_render(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_render.grape') do |*args|
+            endpoint_render(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_run_filters.grape') do |*args|
+            endpoint_run_filters(*args)
+          end
+        end
+
+        def self.endpoint_start_process(*)
+          return if Thread.current[KEY_RUN]
+
+          # retrieve the tracer from the PIN object
+          pin = Datadog::Pin.get_from(::Grape)
+          return unless pin && pin.enabled?
+
+          # store the beginning of a trace
+          tracer = pin.tracer
+          service = pin.service
+          type = Datadog::Ext::HTTP::TYPE
+          tracer.trace('grape.endpoint_run', service: service, span_type: type)
+
+          Thread.current[KEY_RUN] = true
+        rescue StandardError => e
+          Datadog::Tracer.log.error(e.message)
+        end
+
+        def self.endpoint_run(name, start, finish, id, payload)
+          return unless Thread.current[KEY_RUN]
+          Thread.current[KEY_RUN] = false
+
+          # retrieve the tracer from the PIN object
+          pin = Datadog::Pin.get_from(::Grape)
+          return unless pin && pin.enabled?
+
+          tracer = pin.tracer
+          span = tracer.active_span()
+          return unless span
+
+          # collect endpoint details
+          api_view = payload[:endpoint].options[:for].to_s
+          path = payload[:endpoint].options[:path].join('/')
+          resource = "#{api_view}##{path}"
+          span.resource = resource
+
+          # set the request span resource if it's a `rack.request` span
+          request_span = payload[:env][:datadog_rack_request_span]
+          if !request_span.nil? && request_span.name == 'rack.request'
+            request_span.resource = resource
+          end
+
+          # catch thrown exceptions
+          span.set_error(payload[:exception_object]) unless payload[:exception_object].nil?
+
+          # ovverride the current span with this notification values
+          span.start_time = start
+          span.set_tag('grape.route.endpoint', api_view)
+          span.set_tag('grape.route.path', path)
+          span.finish_at(finish)
```

Comment:
This is globally fine. I'm nitpicking but would still probably harden it like in https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/rails/action_controller.rb#L71 by adding an `ensure` clause around `start_time = start` and `finish`. Just, to make sure it makes its way down the pipeline. The heuristic I use here is quite dumb -> if there are at least 10 lines of code, however simple they mike look, the probability it yields an exception "some day in the future" is high.